### ISSUE TITLE
send single ping from ejabberd server

### ISSUE THREAD

**Issue by Veerbhadra277** [processone/ejabberd#1715]
> What version of ejabberd are you using?
Source code Release 16.02

> What operating system (version) are you using?
CentOs7

> How did you install ejabberd (source, package, distribution)?
source

I want to send single ping from ejabberd server (like mod_ping) to specific user, I don't want regular interval pings. is there any way ?

**Comment by zinid** [comment]
There is no particular ejabberdctl command for this as far as I'm concerned.
@badlop may correct me if I'm wrong.

**Comment by Veerbhadra277** [comment]
Thanks @zinid 
Its not about ejabberdctl, I am implementing an erlang module for which need single ping implementation from server to client but it should not be a periodic but there is my own custom logic. is it possible ?

**Comment by zinid** [comment]
Did you try `ejabberd_local:route_iq()`?

**Comment by Veerbhadra277** [comment]
no, I will try this, Thanks @zinid  for your suggestion !

**Comment by Veerbhadra277** [comment]
Thanks @zinid i tried this and i could ping a particular user.
But how could i close session in call back function if i got ping response as undefined ?

**Comment by zinid** [comment]
Is it that hard to read `mod_ping.erl`? :)
TL;DR:
```erlang
ejabberd_c2s:close(Pid, ping_timeout)
```

**Comment by zinid** [comment]
But this is for 17.03+, read the corresponding source code of mod_ping of your ejabberd version, I don't remember the exact API of all ejabberd versions.

**Comment by Veerbhadra277** [comment]
Thanks for reply @zinid  I am using 16.2 source code and seen ejabberd_c2s:close(Pid);  in mod_ping but have some basic doubts in following snippet of my code

` IQ = #iq{type = get,
        sub_el =
        [#xmlel{name = <<"ping">>,
          attrs = [{<<"xmlns">>, ?NS_PING}], children = []}]},

      F = fun(Response) ->
        ?INFO_MSG("Ping response ~p~n",[Response])
          end,

      RecipientJid = jlib:make_jid(Recipient, Server, MobileAppResource),
      ejabberd_local:route_iq(FromRoute, RecipientJid, IQ, F),`

1. Whether callback response function ('F' in above code)  is asynchronous ?
2. If it is asynchronous ? How do i know for which response is for which ping ?

as beginner above query might be very basic but i am trying to implement ping whenever it necessary (as per my business logic) and i want to close session if ping response is timeout.

Regards,
Veer

**Comment by zinid** [comment]
1) Yes, it's asynchronous, it will be called in a separate process
2) The response will be for your exact `IQ`, the respons is matched by `id` (there is no need to set the `id`, ejabberd will to it itself).

**Comment by Veerbhadra277** [comment]
i want to close session if ping response is timeout, but in response I never know for which user ping is timed out, can you please help me in that ?

**Comment by zinid** [comment]
Well, the callback function is a closure, so you can put whatever values into it, e.g.:
```erlang
F = fun(Response) -> ejabberd_c2s:close(Pid) end
```
where Pid can be obtained via `ejabberd_sm:get_session_pid/3` for example.

**Comment by Veerbhadra277** [comment]
For pid we need user ID for which ping is timed out, how we can pass user id in response as callback function is asynchronous?

**Comment by zinid** [comment]
We obtain the Pid **during** request generation, god dammit:
```erlang
Pid = ejabberd_sm:get_session_pid(U, S, R),
F = fun(timeout) -> do_something(Pid); (_) -> ok end,
ejabberd_local:route_iq(From, To, #iq{...}, F)
```
`F` is a **closure**, it copies current arguments on its stack.

**Comment by zinid** [comment]
https://en.wikipedia.org/wiki/Closure_%28computer_programming%29

**Comment by Veerbhadra277** [comment]
@zinid thanks for you help, I will check.
